### DISCUSSION TITLE
Resize render target after render pass ended?

### DISCUSSION THREAD

**Discussion by hjlld** [gpuweb/gpuweb#2275] [Q&A] [answered]
When drawing image on `<canvas>`,  we can do all the other preparing things before `<canvas>`  dom element rendered on the page, except getting canvas's client size and pass it to canvas `width` and `height` attribute.  

```
canvas.width = canvas.clientWidth * window.devicePixelRatio;
canvas.height = canvas.clientHeight * window.devicePixelRatio;
```

This set size operation makes all the other preparing things have to be done after dom rendered. It's a huge waste.

Could `context.configure()` be moved after renderpass ended? Or make possible resize render target after render pass ended but before command buffer submited to device's queue?

And render target's size for canvas `width` and `height` attribute should be defaultly synchronized with its `clientWidth` and `clientHeight` properties? If canvas has not been rendered, could fallback to a fixed size like 300x150.

**Comment by kainino0x** [comment, marked as answer]
In practice, context configuration is not the only thing that has to happen after canvas resize: if you have any other attachments (like a depth buffer), they must also be allocated to a matching size. And in general, resizing textures requires reallocation which is incompatible with the design of WebGPU.

Could you clarify what you mean by "all the other preparing things"? Even in WebGL and 2d canvas, you can't issue draw calls prior to canvas resize, as iirc resizing will wipe the canvas. So perhaps you can provide a clarifying example of what you can achieve in WebGL or 2D canvas that you can't achieve in WebGPU.

In WebGPU, just like in 2d/webgl, if you don't update the drawing buffer width/height\* to match the clientWidth/clientHeight, the content will be stretched to fit. So you'll end up with one frame where the content is stretched, and then you can render at native resolution in the next frame.

(\* in 2d/webgl by updating canvas.width/height; in webgpu by reconfiguring to the appropriate resolution)

In WebGPU, you can pre-record rendering work into a GPURenderBundle and then issue it later. Maybe this helps with your issue?

> And render target's size for canvas `width` and `height` attribute should be defaultly synchronized with its `clientWidth` and `clientHeight` properties? If canvas has not been rendered, could fallback to a fixed size like 300x150.

IIUC, this would be inconsistent with all of the existing canvases (2d and webgl), so we wouldn't make an exception for WebGPU canvases.

**Reply by hjlld** [reply]
Create renderpass depends on its render target, if its render target is a canvas, any preparing job related to the size thing has to wait for the canvas DOM element rendered. In modern frameworks like React, Vue.js, ect... they all use VDOMs, there's plenty of time to do many things before the real DOM rendered but now we cannot do anything but wait during that period. This is a huge waste and what i concerned.

Yes, render bundle is great for this example, but it's only for some static scene, one solution like babylon.js is 1) disable render bundel 2) change the scene 3)re-enable render bundle.

**Reply by Kangz** [reply]
In WebGPU you set the canvas size yourself in `GPUCanvasContext.configure` so you don't need to wait for the canvas to be rendered by the DOM. You just set the size at which you want to render the canvas, and if the size doesn't match the DOM size, then the browser will stretch the content automatically for you.

**Reply by hjlld** [reply]
> In WebGPU you set the canvas size yourself in `GPUCanvasContext.configure` so you don't need to wait for the canvas to be rendered by the DOM. You just set the size at which you want to render the canvas, and if the size doesn't match the DOM size, then the browser will stretch the content automatically for you.

I considered this solution but had not tried, because I thought the aspect ratio is still unknown, browser auto stretch would make distortions.

But when I try it out, obviously whatever the aspect ratio in `context.configure({ size })` is, the final strech doesn't make any distortion. It's great.

**Reply by kainino0x** [reply]
The stretching may change stretch the drawing buffer to a different ratio, but perhaps your actual rendering chooses its aspect ratio based on the DOM size instead of the drawing buffer size? In which case they will cancel out and the final result won't be distorted.

I'll "Mark as answer" but please follow up if you still have issues.
